libtenzir/builtins/operators/print.cpp

```diff
@@ -0,0 +1,198 @@
+
+class print_operator final : public crtp_operator<print_operator> {
+public:
+  print_operator() = default;
+
+  explicit print_operator(parser_interface& p) {
+    const auto* printer = static_cast<const printer_parser_plugin*>(nullptr);
+    try {
+      auto input = p.accept_shell_arg();
+      if (not input) {
+        diagnostic::error("expected extractor")
+          .primary(p.current_span())
+          .throw_();
+      }
+      input_ = std::move(*input);
+      auto printer_name = p.accept_shell_arg();
+      if (not printer_name) {
+        diagnostic::error("expected printer name")
+          .primary(p.current_span())
+          .throw_();
+      }
+      printer_name_ = std::move(*printer_name);
+      printer = plugins::find<printer_parser_plugin>(printer_name_->inner);
+      if (not printer) {
+        diagnostic::error("printer `{}` was not found", printer_name_->inner)
+          .primary(printer_name_->source)
+          .hint("must be one of: {}",
+                fmt::join(std::views::transform(
+                            collect(plugins::get<printer_parser_plugin>()),
+                            &printer_parser_plugin::name),
+                          ", "))
+          .throw_();
+      }
+    } catch (diagnostic& d) {
+      std::move(d)
+        .modify()
+        .usage("print <input> <printer> <args>...")
+        .docs("https://docs.tenzir.com/operators/print")
+        .throw_();
+    }
+    TENZIR_ASSERT(printer);
+    printer_ = printer->parse_printer(p);
+    TENZIR_ASSERT(printer_);
+  }
+
+  auto
+  operator()(generator<table_slice> input, operator_control_plane& ctrl) const
+    -> generator<table_slice> {
+    for (auto&& slice : input) {
+      if (slice.rows() == 0) {
+        co_yield {};
+        continue;
+      }
+      auto target_index
+        = slice.schema().resolve_key_or_concept_once(input_.inner);
+      if (not target_index) {
+        co_return diagnostic::error("could not resolve `{}` for schema `{}`",
+                                    input_.inner, slice.schema())
+          .primary(input_.source)
+          .emit(ctrl.diagnostics());
+      }
+      auto transform = [&](struct record_type::field field,
+                           std::shared_ptr<arrow::Array> array)
+        -> std::vector<
+          std::pair<struct record_type::field, std::shared_ptr<arrow::Array>>> {
+        if (not caf::holds_alternative<record_type>(field.type)) {
+          diagnostic::error("field {} is not of type record", field.name)
+            .primary(input_.source)
+            .throw_();
+        }
+        field.type = type{"dummy_head", field.type};
+        auto rb = arrow::RecordBatch::Make(
+          field.type.to_arrow_schema(), array->length(),
+          static_cast<const arrow::StructArray&>(*array).Flatten().ValueOrDie());
+        auto row = table_slice{rb, field.type};
+        auto builder = series_builder();
+        auto printer_instance = printer_->instantiate(field.type, ctrl);
```

Comment:
After this we need to check whether an error was emitted through the control plane, and if it was, throw it.